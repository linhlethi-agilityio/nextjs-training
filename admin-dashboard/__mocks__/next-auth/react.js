export const useSession = () => ({
  data: {
    user: {
      token: 'access-token',
      email: 'user@gmail.com',
      id: '1',
      name: 'User',
      role: 'admin',
    },
    expires: Date.now().toString(),
  },
  status: 'authenticated',
  update: async () => null,
});
