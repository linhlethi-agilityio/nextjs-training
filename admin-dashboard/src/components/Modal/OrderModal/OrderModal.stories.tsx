import type { Meta, StoryObj } from '@storybook/react';

// Components
import { OrderModal } from '.';

const meta: Meta<typeof OrderModal> = {
  component: OrderModal,
};

export default meta;
type Story = StoryObj<typeof OrderModal>;

export const Default: Story = {
  args: {
    isOpen: true,
    onClose: () => null,
    handleSubmitForm: () => null,
  },
};
