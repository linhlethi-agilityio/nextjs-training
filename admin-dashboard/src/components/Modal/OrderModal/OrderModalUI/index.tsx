'use client';

import { memo, useMemo } from 'react';
import { Controller, useForm } from 'react-hook-form';
import {
  Box,
  Button,
  FormControl,
  FormHelperText,
  FormLabel,
  Input,
} from '@chakra-ui/react';

// Constants
import {
  ERROR_MESSAGES,
  OPTIONS_STATUS,
  REGEX_PATTERN,
  REQUIRED_FIELDS,
} from '@/constants';

// Models
import { Customer, Order } from '@/models';

// Utils
import {
  clearErrorOnChange,
  formatDateString,
  getDirtyState,
  isEnableSubmitButton,
  isValidPrice,
  isValidString,
  normalizeValue,
} from '@/utils';

// Components
import { CustomModal, Dropdown } from '@/components';

interface CustomerModalDetailModalProps {
  title?: string;
  isOpen: boolean;
  isLoading?: boolean;
  customers: Customer[];
  previewData?: Order | null;
  onSubmitForm: (data: Partial<Order>) => void;
  onClose: () => void;
}

const initFormData = {
  product: '',
  customerId: '',
  status: '',
  deadline: '',
  price: undefined,
};

const OrderModalUI = ({
  isOpen,
  customers,
  isLoading = false,
  title = 'Add Order',
  previewData = null,
  onClose,
  onSubmitForm,
}: CustomerModalDetailModalProps) => {
  const formattedOrder = {
    ...previewData,
    deadline: formatDateString(previewData?.deadline ?? '', false, true),
  };

  const optionsCustomers = customers.map((customer) => ({
    value: customer.id,
    label: customer.name,
  }));

  const {
    control,
    formState: { dirtyFields, errors, defaultValues },
    clearErrors,
    handleSubmit,
    watch,
  } = useForm<Order>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    defaultValues: formattedOrder ? formattedOrder : initFormData,
  });

  // Checking to disable/enable submit button
  const dirtyItems = Object.keys(dirtyFields);

  const enableSubmit: boolean = useMemo(
    () => isEnableSubmitButton(REQUIRED_FIELDS, dirtyItems, errors),
    [dirtyItems, errors],
  );

  const isDisableSubmit = !(
    enableSubmit || !getDirtyState(defaultValues ?? {}, watch())
  );

  const handleOnSubmit = (formData: Partial<Order>) => {
    onSubmitForm(formData);
  };

  return (
    <CustomModal isOpen={isOpen} onClose={onClose} title={title} size="3xl">
      <FormControl textAlign="center">
        {/* Name */}
        <Controller
          name="product"
          control={control}
          rules={{
            required: ERROR_MESSAGES.FIELD_REQUIRED,
            validate: (name) =>
              isValidString(name) || ERROR_MESSAGES.INVALID_NAME('product'),
          }}
          render={({
            field: { name, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Box marginBottom={error?.message ? 0 : 25}>
              <FormLabel fontSize="sm">Product</FormLabel>
              <Input
                data-testid="product"
                variant="primary"
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
              {error?.message && (
                <FormHelperText
                  color="red.500"
                  fontSize="xs"
                  mb={1}
                  textAlign="left"
                >
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        <Controller
          name="customerId"
          control={control}
          rules={{
            required: ERROR_MESSAGES.FIELD_REQUIRED,
          }}
          render={({
            field: { name, value, onChange },
            fieldState: { error },
          }) => (
            <Box mb={27}>
              <FormLabel fontSize="sm">Customer</FormLabel>
              <Dropdown
                onChangeValue={(value: string) => {
                  onChange(value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                value={value}
                w="full"
                fontSize={14}
                color="textDark"
                options={optionsCustomers}
                placeholder="Select customers"
              />

              {error?.message && (
                <FormHelperText
                  color="textDanger"
                  fontSize="xs"
                  mb={1}
                  textAlign="left"
                >
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        <Controller
          name="status"
          control={control}
          rules={{
            required: ERROR_MESSAGES.FIELD_REQUIRED,
          }}
          render={({
            field: { name, value, onChange },
            fieldState: { error },
          }) => (
            <Box mb={21}>
              <FormLabel fontSize="sm">Status</FormLabel>
              <Dropdown
                onChangeValue={(value: string) => {
                  onChange(value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                value={value}
                w="full"
                fontSize={14}
                color="textDark"
                options={OPTIONS_STATUS}
                placeholder="Select status"
              />

              {error?.message && (
                <FormHelperText
                  color="red.500"
                  fontSize="xs"
                  mb={1}
                  textAlign="left"
                >
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        <Controller
          name="deadline"
          control={control}
          rules={{ required: ERROR_MESSAGES.FIELD_REQUIRED }}
          render={({
            field: { name, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Box marginBottom={error?.message ? 0 : 25}>
              <FormLabel fontSize="sm">Deadline</FormLabel>
              <Input
                color="textDark"
                data-testid="deadline"
                variant="primary"
                type="date"
                min={new Date().toISOString().split('T')[0]}
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
              {error?.message && (
                <FormHelperText
                  color="red.500"
                  fontSize="xs"
                  mb={1}
                  textAlign="left"
                >
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        <Controller
          name="price"
          control={control}
          rules={{
            required: ERROR_MESSAGES.FIELD_REQUIRED,
            validate: (value) =>
              isValidPrice(value!.toString()) || ERROR_MESSAGES.INVALID_PRICE,
          }}
          render={({
            field: { name, value, onChange, ...rest },
            fieldState: { error },
          }) => (
            <Box marginBottom={error?.message ? 0 : 25}>
              <FormLabel fontSize="sm">Price</FormLabel>
              <Input
                data-testid="price"
                value={value || ''}
                variant="primary"
                isInvalid={!!error?.message}
                onChange={(e) => {
                  const value = normalizeValue(e.target.value);

                  const isValidate = REGEX_PATTERN.DECIMAL.test(value);

                  if (isValidate || value === '') onChange(value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
              {error?.message && (
                <FormHelperText
                  color="red.500"
                  fontSize="xs"
                  mb={1}
                  textAlign="left"
                >
                  {error.message}
                </FormHelperText>
              )}
            </Box>
          )}
        />

        <Button
          isLoading={isLoading}
          colorScheme="brand"
          aria-label="button-submit"
          type="submit"
          w="full"
          isDisabled={isDisableSubmit}
          marginBottom={5}
          onClick={handleSubmit(handleOnSubmit)}
        >
          Submit
        </Button>
      </FormControl>
    </CustomModal>
  );
};

export default memo(OrderModalUI);
