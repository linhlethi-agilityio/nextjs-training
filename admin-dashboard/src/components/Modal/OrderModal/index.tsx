// Models
import { Order } from '@/models';

// Components
import OrderModalUI from './OrderModalUI';

// Api
import { getTotalCustomers } from '@/api';

interface CustomerModalDetailModalProps {
  title?: string;
  isOpen: boolean;
  isLoading?: boolean;
  previewData?: Order | null;
  onSubmitForm: (data: Partial<Order>) => void;
  onClose: () => void;
}

const OrderModal = async ({
  title = 'Add Order',
  isLoading = false,
  isOpen,
  previewData,
  onClose,
  onSubmitForm,
}: CustomerModalDetailModalProps) => {
  return (
    <OrderModalUI
      isOpen={isOpen}
      isLoading={isLoading}
      title={title}
      previewData={previewData}
      onClose={onClose}
      getTotalCustomers={getTotalCustomers}
      onSubmitForm={onSubmitForm}
    />
  );
};

export default OrderModal;
