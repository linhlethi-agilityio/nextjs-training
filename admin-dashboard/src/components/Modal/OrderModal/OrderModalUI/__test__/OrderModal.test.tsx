import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import OrderModal from '..';

const customers = [
  {
    createdAt: '2024-07-21T00:07:42.462Z',
    name: 'Connie Smitham',
    email: 'Ansley_Jast7@hotmail.com',
    phoneNumber: '566-231-4136',
    dateOfBirth: '1979-11-21T13:46:01.346Z',
    shippingAddress: 'East',
    id: '1',
  },
  {
    createdAt: '2024-07-21T05:43:59.000Z',
    name: 'Terence Murphy',
    email: 'Johnson_Mayer@hotmail.com',
    phoneNumber: '263-583-0766',
    dateOfBirth: '1966-05-03T08:57:27.517Z',
    shippingAddress: 'South',
    id: '2',
  },
];

// Mock handleOnChange function
const mockOnSubmitForm = jest.fn();
const mockOnclose = jest.fn();

const mockProps = {
  title: 'Add order',
  isOpen: true,
  onClose: mockOnclose,
  onSubmitForm: mockOnSubmitForm,
  getTotalCustomers: jest.fn(() =>
    Promise.resolve({
      data: customers,
    }),
  ),
};

describe('OrderModal component', () => {
  it('should match snapshot for OrderModal', () => {
    const { container } = render(<OrderModal {...mockProps} />);

    expect(container).toMatchSnapshot();
  });
});
