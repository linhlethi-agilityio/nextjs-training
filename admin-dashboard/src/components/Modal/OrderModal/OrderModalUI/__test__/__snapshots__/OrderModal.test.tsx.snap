// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderModal component should match snapshot for OrderModal 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
